### PR TITLE
Refactor platform dependent gem creation.

### PR DESCRIPTION
**Pull request by ksubrama**
Since we don't actually have platform specific code, only dependencies, we can use this structure to generate gems based on just the ABI needed.  Hence we have a mingw32 package (supports both 32 and 64-bit outputs) and an mswin32 package (good luck fellas..).  We also have a pure ruby output for test purposes.

### CODE REVIEW

**Comment by ksubrama** [comment]
@jdmundrawala @btm @adamedx

File: Rakefile [hunk 1]
```diff
@@ -14,7 +46,7 @@ RSpec::Core::RakeTask.new(:spec) do |t|
 end
 
 desc "Build it and ship it"
-task :ship => [:clean, :gem] do
+task ship: :gems do
```
Comment (jdmundrawala) [inline Rakefile]: is this syntax correct?

File: lib/mixlib/shellout/version.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Mixlib
   class ShellOut
-    VERSION = "2.0.1"
+    VERSION = "2.1.0"
```
Comment (jdmundrawala) [inline lib/mixlib/shellout/version.rb]: This isn't necessary for this PR

File: mixlib-shellout-x86-mingw32.gemspec [hunk 3]
```diff
@@ -1,9 +0,0 @@
-# x86-mingw32 Gemspec #
```
Comment (jdmundrawala) [inline mixlib-shellout-x86-mingw32.gemspec]: Looks like bundler is going to want this file

Comment (ksubrama) [inline mixlib-shellout-x86-mingw32.gemspec]: Blargh...  I tried to get bundler to conditionally recognize/install platforms but that seems to not be currently supported.  Either that or I seem to not understand all the interdependencies between rake and gem and bundler.

**Comment by ksubrama** [comment]
I made some changes and squashed.  We now have 1 universal package on windows and one "ruby" platform package.  This should hopefully work better.  /cc @jdmundrawala

File: Rakefile [hunk 4]
```diff
@@ -1,20 +1,19 @@
 require 'rspec/core/rake_task'
 require 'rubygems/package_task'
+require 'mixlib/shellout/version'
 
-Dir[File.expand_path("../*gemspec", __FILE__)].reverse.each do |gemspec_path|
-  gemspec = eval(IO.read(gemspec_path))
+Dir[File.expand_path("../*gemspec", __FILE__)].each do |gemspec_path|
```
Comment (jdmundrawala) [inline Rakefile]: Why this change?

File: Rakefile [hunk 5]
```diff
@@ -1,20 +1,19 @@
...
-require 'mixlib/shellout/version'
-
 desc "Run all specs in spec directory"
 RSpec::Core::RakeTask.new(:spec) do |t|
   t.pattern = FileList['spec/**/*_spec.rb']
 end
 
 desc "Build it and ship it"
-task :ship => [:clean, :gem] do
+task ship: [:clobber_package, :gem] do
```
Comment (jdmundrawala) [inline Rakefile]: why clobber_package ver clean?

File: Rakefile (hunk 4 above)
Comment (ksubrama) [inline Rakefile]: Was unsure why it was needed - the commit log doesn't seem to mention needing to build this in any specific order.  This was mostly a "I wonder what would break" change and so far, it doesn't seem to break anything.

Comment (jdmundrawala) [inline Rakefile]: I'd rather not find out. It's unrelated to what you are trying to accomplish

File: Rakefile (hunk 5 above)
Comment (ksubrama) [inline Rakefile]: The clean task was actually not available on my machine.  Don't know if this was a recent change or what.  :gem seems to add the cleaning directives to :clobber_package - so I added that dependency instead.

File: Rakefile (hunk 4 above)
Comment (ksubrama) [inline Rakefile]: :) Fixed.

File: Rakefile (hunk 5 above)
Comment (jdmundrawala) [inline Rakefile]: Did you `bundle exec` rake?

Comment (ksubrama) [inline Rakefile]: Yep.

C:\Users\ksubramanian\repo\chef\mixlib-shellout [ksubrama/win64]> bundle exec rake clean
rake aborted!
Don't know how to build task 'clean'

(See full trace by running task with --trace)
C:\Users\ksubramanian\repo\chef\mixlib-shellout [ksubrama/win64]> bundle exec rake -T
rake clobber_package  # Remove package products
rake gem              # Build the gem file mixlib-shellout-2.0.1.gem / Buil...
rake package          # Build all the packages
rake repackage        # Force a rebuild of the package files
rake ship             # Build it and ship it
rake spec             # Run all specs in spec directory

There is no Gemfile.lock in this repo - so it just grabs a recent rake.

Comment (jdmundrawala) [inline Rakefile]: `require 'rake/clean'`?

Comment (ksubrama) [inline Rakefile]: I had gone down that route before.  rake/clean didn't get its clean targets from the Gem tasks.  Hence, running rake clean did nothing.  To get that to work, I had to manually add 'pkg' to CLEANS or just define my own clean => :clobber_package task.

**Comment by jdmundrawala** [comment]
:+1:

File: mixlib-shellout-x86-mingw32.gemspec (hunk 3 above)
Comment (btm) [inline mixlib-shellout-x86-mingw32.gemspec]: @ksubrama Did you figure this out?

**Comment by jdmundrawala** [comment]
I take back my :+1: until it works with bundler

File: mixlib-shellout-x86-mingw32.gemspec (hunk 3 above)
Comment (ksubrama) [inline mixlib-shellout-x86-mingw32.gemspec]: Yes I did - I was confused about how bundler was actually handling gemspecs.  Turns out that you can only point it to a single gemspec file in your Gemfile, but then what it does (either by design or otherwise) is that it loads all gemfiles located at that path regardless of the file name.  It then looks for all gemspec objects whose name attribute corresponds to the name provided in the Gemfile, filters out all gemspec objects that don't match the current platform and then "sorts" the list (I haven't figured out exactly how it sorts it) and picks the top match.  It's convoluted but this works with it.  It installs windows-y things on windows and posix-y things otherwise.

**Comment by ksubrama** [comment]
It does :)
